lib/ecto/multi.ex

```diff
@@ -0,0 +1,364 @@
+      iex> Ecto.Multi.prepend(lhs, rhs) |> Ecto.Multi.to_list |> Keyword.keys
+      [:right, :left]
+
+  """
+  @spec prepend(t, t) :: t
+  def prepend(lhs, rhs) do
+    merge(lhs, rhs, &(&1 ++ &2))
+  end
+
+  defp merge(%Multi{} = lhs, %Multi{} = rhs, joiner) do
+    %{names: lhs_names, operations: lhs_ops} = lhs
+    %{names: rhs_names, operations: rhs_ops} = rhs
+    if MapSet.disjoint?(lhs_names, rhs_names) do
+      %Multi{names: MapSet.union(lhs_names, rhs_names),
+             operations: joiner.(lhs_ops, rhs_ops)}
+    else
+      common = MapSet.intersection(lhs_names, rhs_names) |> MapSet.to_list
+      raise """
+      when merging following Ecto.Multi:
+
+      #{inspect lhs}
+
+      #{inspect rhs}
+
+      both declared operations: #{inspect common}
+      """
+    end
+  end
+
+  @doc """
+  Adds an insert operation to the multi.
+
+  Accepts the same arguments and options as `Ecto.Repo.insert/3` does.
+  """
+  @spec insert(t, name, Changeset.t | Ecto.Schema.t, Keyword.t) :: t
+  def insert(multi, name, changeset_or_struct, opts \\ [])
+
+  def insert(multi, name, %Changeset{} = changeset, opts) do
+    add_changeset(multi, :insert, name, changeset, opts)
+  end
+
+  def insert(multi, name, struct, opts) do
+    insert(multi, name, Changeset.change(struct), opts)
+  end
+
+  @doc """
+  Adds an update operation to the multi.
+
+  Accepts the same arguments and options as `Ecto.Repo.update/3` does.
+  """
+  @spec update(t, name, Changeset.t, Keyword.t) :: t
+  def update(multi, name, %Changeset{} = changeset, opts \\ []) do
+    add_changeset(multi, :update, name, changeset, opts)
+  end
+
+  @doc """
+  Adds a delete operation to the multi.
+
+  Accepts the same arguments and options as `Ecto.Repo.delete/3` does.
+  """
+  @spec delete(t, name, Changeset.t | Ecto.Schema.t, Keyword.t) :: t
+  def delete(multi, name, changeset_or_struct, opts \\ [])
+
+  def delete(multi, name, %Changeset{} = changeset, opts) do
+    add_changeset(multi, :delete, name, changeset, opts)
+  end
+
+  def delete(multi, name, struct, opts) do
+    delete(multi, name, Changeset.change(struct), opts)
+  end
+
+  defp add_changeset(multi, action, name, changeset, opts) when is_list(opts) do
+    add_operation(multi, name, {:changeset, put_action(changeset, action), opts})
+  end
+
+  defp put_action(%{action: nil} = changeset, action) do
+    %{changeset | action: action}
+  end
+
+  defp put_action(%{action: action} = changeset, action) do
+    changeset
+  end
+
+  defp put_action(%{action: original}, action) do
+    raise ArgumentError, "you provided a changeset with an action already set " <>
+      "to #{inspect original} when trying to #{action} it"
+  end
+
+  @doc """
+  Adds a function to run as part of the multi
+
+  The function should return either `{:ok, value}` or `{:error, value}`, and
+  receives changes so far as an argument.
+  """
+  @spec run(t, name, run) :: t
+  def run(multi, name, run) when is_function(run, 1) do
```

Comment:
`&charge_client(&1, "product_id")` will produce an anonymous function of arity 1 since it accepts only one additional argument. 

We'll always call this function with one argument - changes so far, so we should blow up early if given wrong functions.